### PR TITLE
AzureFirewall with IpGroups (LATEST)

### PR DESCRIPTION
**Pull request by ssripadham**
## Description

    - `Azure Firewall will support IpGroups can be used in the network/application/dnat rules as described below.

- As a source or destination address in AZFW network rules
- As a source address in AZFW application rules
- As a source address in DNAT rules. `

**1. Azure Firewall Application Rules**

         "AzureFirewallApplicationRule": {
             "properties": {
               "name": {
                 "type": "string",
                 "description": "Name of the application rule."
               },
               "description": {
                 "type": "string",
                 "description": "Description of the rule."
               },
               ...
               "sourceIpGroups": {
                 "type": "array",
                 "description": "List of source IP Groups for this rule.",
                 "items": {
                   "type": "string"
                 }
               },
           ...
         }
        }

**2. Azure Firewall  NAT Rules**

       "AzureFirewallNatRule": {
             "properties": {
               "name": {
                 "type": "string",
                 "description": "Name of the NAT rule."
               },
               "description": {
                 "type": "string",
                 "description": "Description of the rule."
               },
              ...

        "sourceIpGroups": {
          "type": "array",
          "description": "List of source IP Groups for this rule.",
          "items": {
            "type": "string"
          }
        }
           …
       }
       }

**3. Azure Firewall  Network Rules**

        "AzureFirewallNetworkRule": {
             "properties": {
               "name": {
                 "type": "string",
                 "description": "Name of the network rule."
               },
               "description": {
                 "type": "string",
                 "description": "Description of the rule."
               },
              ...
               "sourceIpGroups": {
                 "type": "array",
                 "description": "List of source IP Groups for this rule.",
                 "items": {
                   "type": "string"
                 }
               },
               "destinationIpGroups": {
                 "type": "array",
                 "description": "List of destination IP Groups for this rule.",
                 "items": {
                   "type": "string"
                 }
               },
           ...
        }
       }

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Network/Network/AzureFirewall/ApplicationRule/NewAzureFirewallApplicationRuleCommand.cs [hunk 1]
```diff
@@ -27,7 +27,7 @@ public class NewAzureFirewallApplicationRuleCommand : AzureFirewallBaseCmdlet
             Mandatory = true,
             ParameterSetName = AzureFirewallApplicationRuleParameterSets.TargetFqdn,
             HelpMessage = "The name of the Application Rule")]
-        [Parameter(
+         [Parameter(
```
Comment (VeryEarly) [inline src/Network/Network/AzureFirewall/ApplicationRule/NewAzureFirewallApplicationRuleCommand.cs]: remove the white space

File: src/Network/Network/AzureFirewall/NatRule/NewAzureFirewallNatRuleCommand.cs [hunk 2]
```diff
@@ -90,6 +94,12 @@ public override void Execute()
             // Add some validation based on the type of RuleCollection (SNAT will be supported later)
             // if (MNM.AzureFirewallNatRCActionType.Dnat.Equals(ActionType))
             {
+               // One of SourceAddress or SourceIpGroup must be present
+               if ((SourceAddress == null) && (SourceIpGroup == null))
```
Comment (VeryEarly) [inline src/Network/Network/AzureFirewall/NatRule/NewAzureFirewallNatRuleCommand.cs]: it looks like you need two parameter sets,  SourceAddress and SourceIpGroup will need to be mandatory for each.

File: src/Network/Network/AzureFirewall/NetworkRule/NewAzureFirewallNetworkRuleCommand.cs [hunk 3]
```diff
@@ -76,6 +85,12 @@ public override void Execute()
         {
             base.Execute();
 
+            // One of SourceAddress or SourceIpGroup must be present
+            if ((SourceAddress == null) && (SourceIpGroup == null))
```
Comment (VeryEarly) [inline src/Network/Network/AzureFirewall/NetworkRule/NewAzureFirewallNetworkRuleCommand.cs]: SourceAddress is mandatory, you are trying to throw exception when SourceIpGroup is null, are you trying to have another parameter set which requires SourceIpGroup mandatory?

File: src/Network/Network/AzureFirewall/NatRule/NewAzureFirewallNatRuleCommand.cs (hunk 2 above)
Comment (ssripadham) [reply src/Network/Network/AzureFirewall/NatRule/NewAzureFirewallNatRuleCommand.cs]: Fixed to make SourceAddress optional. This check is to flag when both fields are optional

File: src/Network/Network/AzureFirewall/ApplicationRule/NewAzureFirewallApplicationRuleCommand.cs (hunk 1 above)
Comment (ssripadham) [reply src/Network/Network/AzureFirewall/ApplicationRule/NewAzureFirewallApplicationRuleCommand.cs]: fixed

File: src/Network/Network/AzureFirewall/NatRule/NewAzureFirewallNatRuleCommand.cs (hunk 2 above)
Comment (ssripadham) [reply src/Network/Network/AzureFirewall/NatRule/NewAzureFirewallNatRuleCommand.cs]: No, we do not want to use the parameter set. It lead us to complication in paramterset definition as in https://github.com/Azure/azure-powershell/pull/10668/files
We are using an approach where both SourceAddress  & SourceIpGroup are optional and using checks to make sure atleast one of these is supplied by the user

File: src/Network/Network/AzureFirewall/NetworkRule/NewAzureFirewallNetworkRuleCommand.cs (hunk 3 above)
Comment (ssripadham) [reply src/Network/Network/AzureFirewall/NetworkRule/NewAzureFirewallNetworkRuleCommand.cs]: Fixed to make SourceAddress optional. This check is to flag when both fields are null

File: src/Network/Network/AzureFirewall/ApplicationRule/NewAzureFirewallApplicationRuleCommand.cs (hunk 1 above)
Comment (ssripadham) [reply src/Network/Network/AzureFirewall/ApplicationRule/NewAzureFirewallApplicationRuleCommand.cs]: Done

File: src/Network/Network/AzureFirewall/NatRule/NewAzureFirewallNatRuleCommand.cs (hunk 2 above)
Comment (ssripadham) [reply src/Network/Network/AzureFirewall/NatRule/NewAzureFirewallNatRuleCommand.cs]: We are using similar approach for Destination Addresses in network rule.

Comment (VeryEarly) [reply src/Network/Network/AzureFirewall/NatRule/NewAzureFirewallNatRuleCommand.cs]: the similar approach is not appropriate either, breaking change fix needed

File: src/Network/Network/AzureFirewall/NatRule/NewAzureFirewallNatRuleCommand.cs [hunk 4]
```diff
@@ -90,6 +94,12 @@ public override void Execute()
...
+               {
+                  throw new ArgumentException("Either SourceAddress or SourceIpGroup is required.");
```
Comment (VeryEarly) [inline src/Network/Network/AzureFirewall/NatRule/NewAzureFirewallNatRuleCommand.cs]: better to put exception message here https://github.com/Azure/azure-powershell/blob/ipgroup_azfw_branch/src/Network/Network/Properties/Resources.resx

File: src/Network/Network/AzureFirewall/NetworkRule/NewAzureFirewallNetworkRuleCommand.cs [hunk 5]
```diff
@@ -87,13 +102,13 @@ public override void Execute()
             // Only one of DestinationAddress or DestinationFqdns is allowed
             if ((DestinationAddress != null) && (DestinationFqdn != null))
             {
-                throw new ArgumentException("Both DestinationAddress and DestinationFqdns not allowed");
+                throw new ArgumentException("Both DestinationAddress and DestinationFqdns not allowed.");
```
Comment (VeryEarly) [inline src/Network/Network/AzureFirewall/NetworkRule/NewAzureFirewallNetworkRuleCommand.cs]: I understand you didn't create this message but this line is confusing, can you put something like "only one of DestinationAddress or DestinationFqdns is allowed"?

**Pull request merged**